metrics-sampler/src/churn_metrics_sampler.rs

```diff
@@ -0,0 +1,202 @@
+//! Samples churn-related metrics at a fixed interval
+
+use std::{
+    collections::{HashMap, HashSet},
+    time::Duration,
+};
+
+use common::types::{
+    exchange::PriceReporterState, network_order::NetworkOrderState, token::Token,
+    wallet::order_metadata::OrderMetadata, Price,
+};
+use futures::future::join_all;
+use job_types::price_reporter::{PriceReporterJob, PriceReporterQueue};
+use num_bigint::BigUint;
+use state::State;
+use util::hex::biguint_to_hex_addr;
+
+use crate::sampler::AsyncMetricSampler;
+
+/// A mapping from token pair to price
+type PriceMap = HashMap<(Token, Token), Price>;
+
+/// The name of the sampler for churn metrics
+const CHURN_METRICS_SAMPLER_NAME: &str = "churn-metrics-sampler";
+/// The interval at which to sample churn metrics
+const CHURN_METRICS_SAMPLE_INTERVAL_MS: u64 = 3_600_000; // 1 hour
+
+/// Metric describing the number of orders cancelled
+const NUM_ORDERS_CANCELLED_METRIC: &str = "num_orders_cancelled";
+/// Metric describing the value of a cancelled order
+const CANCELLED_ORDER_VALUE_METRIC: &str = "cancelled_order_value";
+/// Metric describing the % filled of a cancelled order
+const CANCELLED_ORDER_FILL_PERCENT_METRIC: &str = "cancelled_order_fill_percent";
+
+/// Metric tag denoting the order ID
+const ORDER_ID_METRIC_TAG: &str = "order_id";
+/// Metric tag denoting the base ticker
+const BASE_TICKER_METRIC_TAG: &str = "base_ticker";
+/// Metric tag denoting the quote ticker
+const QUOTE_TICKER_METRIC_TAG: &str = "quote_ticker";
+
+/// Samples churn metrics at a fixed interval
+#[derive(Clone)]
+pub struct ChurnMetricsSampler {
+    /// A handle to the global state
+    state: State,
+    /// A handle to the price reporter's job queue
+    price_reporter_job_queue: PriceReporterQueue,
+}
+
+impl ChurnMetricsSampler {
+    /// Create a new `ChurnMetricsSampler`
+    pub fn new(state: State, price_reporter_job_queue: PriceReporterQueue) -> Self {
+        Self { state, price_reporter_job_queue }
+    }
+
+    // -------------------
+    // | Private Helpers |
+    // -------------------
+
+    /// Get the metadata for all cancelled orders
+    async fn get_cancelled_orders(&self) -> Result<Vec<OrderMetadata>, String> {
+        let orders = self.state.get_all_orders().await?;
+        let cancelled_orders =
+            orders.into_iter().filter(|o| matches!(o.state, NetworkOrderState::Cancelled));
+
+        let cancelled_orders_meta = join_all(
+            cancelled_orders.map(|o| async move { self.state.get_order_metadata(&o.id).await }),
+        )
+        .await
+        .into_iter()
+        .filter_map(Result::ok)
+        .map(|o| o.unwrap())
+        .collect();
+
+        Ok(cancelled_orders_meta)
+    }
+
+    /// Get the prices for all unique pairs across the given orders
+    async fn get_order_prices(
```

Comment:
Actually, I'd just get rid of this method and re-fetch a price for each order. it takes no time to fetch a price (on the order of 2-digit nanoseconds), prefetching and aggregating them for an already expensive operation we do will make little impact on the execution time here. It just makes things harder to read to pass around a hashmap

Reply:
fair enough, happy to make this change